lib/ansible/modules/cloud/amazon/aws_step_functions_state_machine_execution.py

```diff
@@ -0,0 +1,178 @@
+start_date:
+    description: The date the execution is started.
+    type: str
+    returned: if action == start and changed == True
+    sample: "2019-11-02T22:39:49.071000-07:00"
+stop_date:
+    description: The date the execution is stopped.
+    type: str
+    returned: if action == stop
+    sample: "2019-11-02T22:39:49.071000-07:00"
+'''
+
+
+from ansible.module_utils.aws.core import AnsibleAWSModule
+
+try:
+    from botocore.exceptions import ClientError, BotoCoreError
+except ImportError:
+    pass  # caught by AnsibleAWSModule
+
+
+def start_execution(module, sfn_client):
+    check_mode(module, msg='State machine execution would be started.', changed=True)
+
+    state_machine_arn = module.params.get('state_machine_arn')
+    name = module.params.get('name')
+    execution_input = module.params.get('execution_input')
+
+    try:
+        execution = sfn_client.start_execution(
+            stateMachineArn=state_machine_arn,
+            name=name,
+            input=execution_input
+        )
+    except ClientError as e:
+        if e.response['Error']['Code'] == 'ExecutionAlreadyExists':
+            module.exit_json(changed=False)
+        module.fail_json_aws(e, msg="Failed to start execution.")
+
+    module.exit_json(changed=True, execution_arn=execution.get('executionArn'),
+                     start_date=execution.get('startDate'))
+
+
+def stop_execution(module, sfn_client):
+    check_mode(module, msg='State machine execution would be stopped.', changed=True)
+
+    cause = module.params.get('cause')
+    error = module.params.get('error')
+    execution_arn = module.params.get('execution_arn')
+
+    try:
+        res = sfn_client.stop_execution(
+            executionArn=execution_arn,
+            cause=cause,
+            error=error
+        )
+    except ClientError as e:
+        module.fail_json_aws(e, msg="Failed to stop execution.")
+
+    module.exit_json(changed=True, stop_date=res.get('stopDate'))
+
+
+def check_mode(module, msg='', changed=False):
+    if module.check_mode:
+        module.exit_json(changed=changed, output=msg)
+
+
+def main():
+    module_args = dict(
+        action=dict(choices=['start', 'stop'], default='start'),
+        name=dict(type='str'),
+        execution_input=dict(type='json', default={}),
+        state_machine_arn=dict(type='str'),
+        cause=dict(type='str', default=''),
+        error=dict(type='str', default=''),
+        execution_arn=dict(type='str')
+    )
+    module = AnsibleAWSModule(
+        argument_spec=module_args,
+        required_if=[('action', 'start', ['name', 'state_machine_arn']),
+                     ('action', 'stop', ['execution_arn']),
+                     ],
+        supports_check_mode=True
+    )
+
+    sfn_client = module.client('stepfunctions')
+
+    action = module.params.get('action')
+    if action == "start":
+        start_execution(module, sfn_client)
```

Comment:
`client.start_execution` is idempotent but `start_execution` always returns `changed=True` if `action=start`. We probably want to `client.list_executions` to see if the execution is already running before proceeding.

Reply:
> The `list_executions` is eventually consistent. The results are best effort and may not reflect very recent updates and changes.

So we could potentially get false positives (changed=True) if the execution was already running but did not get listed by the `list_executions` call.

Reply:
> StartExecution is idempotent. If StartExecution is called with the same name and input as a running execution, the call will succeed and return the same response as the original request. If the execution is closed or if the input is different, it will return a 400 ExecutionAlreadyExists error. Names can be reused after 90 days.

I return `changed=False` if ExecutionAlreadyExists is returned. That would mean that a new execution was not started. 

But I am not sure if we want to return `changed=False` in the case of a running execution. I guess that really depends on how the value of `changed` is going to be used further down the playbook. Also, if the AWS service isn't making the distinction clear maybe we don't need to either. What do you think?

Reply:
I haven't actually used stepfunctions myself so I'm just going off the docs. AIUI, `ExecutionAlreadyExists` would only be returned on a running execution if the `execution_input` is different, not in a case where the input is the same.  As an ansible user who is new to step functions, I feel like if I gave the same input and ran the playbook twice I would expect the 2nd time to be `changed=false` if the first execution is still running.  But that doesn't mean that's how AWS users familiar with step functions will expect it.

Reply:
We could just document that the value of `changed` for the tasks will be aligned to how boto3 call returns. If we get `ExecutionAlreadyExists`, we return `False`. Otherwise we return `True`. Again, what the right value for `changed` is depends on how the user of the module wishes to use it. We should just document it precisely and then gather feedback from the users as they start using the module. Almost wish there was a thing like beta modules to gather early feedback.

Reply:
Between ansible and AWS, we should be more aligned with AWS on this one, IMO. If they don't make a distinction between `ExecutionAlreadyExists` and `ExecutionCurrentlyRunning`, then we shouldn't either. This and the fact that `list_executions` is only eventually consistent makes me think the current behavior is what we want.

Reply:
That makes sense, and I definitely agree we should very clearly document the behaviour.
@tremble, any thoughts on this approach to idempotency here before we proceed?

Reply:
I have a *slight* preference for @jillr's approach, and trying to ensure that check_mode behaves as you might expect.  But it's not a strong enough preference to hold anything up.  The behaviour should certainly be documented though.

Reply:
So...? What should we do here? Document the behavior clearly? If that is what we plan to do I can push another patch.
Given that with the eventually consistent `list_executions` you could get inconsistent results, do we still want to go with Ansible's definition for idempotency here?

Reply:
I like @jillr's approach. We should err on the side of trying to keep our AWS modules similar. There are a lot of AWS services that use eventual consistency so we have to work around it as best we can. Using waiters after making API calls can help a little bit, though I'm not sure that it's applicable here (I see that there are no boto3 built-in waiters yet for this service, but you could possible add your own criteria for waiting to https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/waiters.py). In some circumstances we've added brief pause tasks to integration tests to help. Do add a note in the documentation about the eventual consistency problem.

Reply:
I have pushed more changes. For start execution, we now list all executions and check if there is an execution with the same name. If there is one, the module exits with `changed=False`. For stop execution, we describe the execution and exit with `changed=False` unless the execution status is '_RUNNING_'. I have also added tests around these changes and comments wherever I felt necessary.